numenc/encoder_decoder.c

```diff
@@ -0,0 +1,947 @@
+    PyObject * output;
+    PyObject * input_obj;
+    if (!PyArg_ParseTuple(args, "O", & input_obj)) {
+        return PyErr_Format(PyExc_TypeError, "Wrong input: expected py object.");
+    }
+
+    unsigned long long input = PyLong_AsUnsignedLongLong(input_obj);
+    if (input == -1 && PyErr_Occurred()) {
+        PyErr_Clear();
+        return PyErr_Format(PyExc_TypeError, "Parameter must be an unsigned integer type, but got %s", Py_TYPE(input_obj) -> tp_name);
+    }
+
+    if (endianness == -1) {
+        endianness = endianness_impl();
+    }
+
+    char * buffer = (char * ) PyMem_RawMalloc(EIGHT_BYTES);
+    if (buffer == NULL) {
+        return PyErr_NoMemory();
+    }
+
+    if (endianness == 0) {
+        for (int i = 0; i < EIGHT_BYTES; i++) {
+            buffer[i] = (input >> (i * 8)) & 0xff;
+        }
+    } else {
+        for (int i = 0; i < EIGHT_BYTES; i++) {
+            buffer[EIGHT_BYTES - 1 - i] = (input >> (i * 8)) & 0xff;
+        }
+    }
+
+    output = PyBytes_FromStringAndSize(buffer, EIGHT_BYTES);
+    PyMem_RawFree(buffer);
+    return output;
+}
+
+static PyObject * to_uint64(PyObject * self, PyObject * args) {
+    const char * input;
+    PyObject * output;
+    int count;
+
+    if (!PyArg_ParseTuple(args, "y#", & input, & count)) {
+        return PyErr_Format(PyExc_TypeError, "Wrong input: expected bytes, but got %s", Py_TYPE(args) -> tp_name);
+    }
+
+    if (endianness == -1) {
+        endianness = endianness_impl();
+    }
+
+    if (count != EIGHT_BYTES) {
+        return PyErr_Format(PyExc_ValueError, "Illegal input: expected bytes of length %d, got %d.",
+                            EIGHT_BYTES, count);
+    }
+
+    unsigned char * buffer = (unsigned char * ) PyMem_RawMalloc(EIGHT_BYTES);
+    if (buffer == NULL) {
+        return PyErr_NoMemory();
+    }
+
+    for (int i = 0; i < EIGHT_BYTES; i++) {
+        buffer[i] = (unsigned char) input[i];
+    }
+
+    uint64_t decoded = 0;
+    if (endianness == 0) {
+        for (int i = 0; i < EIGHT_BYTES; i++) {
+            decoded |= ((uint64_t) buffer[i]) << (i * 8);
+        }
+    } else {
+        for (int i = 0; i < EIGHT_BYTES; i++) {
+            decoded |= ((uint64_t) buffer[EIGHT_BYTES - 1 - i]) << (i * 8);
+        }
+    }
+    output = PyLong_FromUnsignedLongLong(decoded);
+    PyMem_RawFree(buffer);
+    return output;
+}
+
+static PyObject * from_float32(PyObject * self, PyObject * args) {
+    float input;
+    PyObject * output;
+
+    if (!PyArg_ParseTuple(args, "f", & input)) {
+        return PyErr_Format(PyExc_TypeError, "Wrong input: expected 32-bit float.");
+    }
+
+    if (endianness == -1) {
+        endianness = endianness_impl();
+    }
+
+    union float_bytes f_bytes;
+    f_bytes.val = input;
+
+    char * buffer = (char * ) PyMem_RawMalloc(FOUR_BYTES);
+    if (buffer == NULL) {
+        return PyErr_NoMemory();
+    }
+
+    // positive number: set sign bit to 1
```

Comment:
(Comment B) I'd put it in the if-body instead of before -- it makes me read this as if this was the condition that you test, not the behavior executed on positive numbers.

The same applies to the comment to `else` below.